### PR TITLE
Code style etc

### LINKED ISSUE #20: Code Style
As we discussed today on the call, we should try and converge on a coding style.

There are a few popular options that I found:

- [black](https://github.com/psf/black)
- [pycodestyle](https://github.com/pycqa/pycodestyle) (formerly pep8)
- [yapf](https://github.com/google/yapf)

----

My personal preference is for black.  We use it on Flux, Spack, and a few other LLNL projects.  It is actually the formatter/style recommended internally at Livermore Computing.  Everyone has a different qualm or nitpick with the style that it enforces, but most people I talk to are fine with it overall.  The major benefit with it is that it has no configuration, so you never spend any time fighting with people over how things should be; they are just what they are.  Another minor benefit is that black makes code review easier my optimizing for smaller diffs.  Relevant excerpt from their README:

> Black is the uncompromising Python code formatter. By using it, you agree to cede control over minutiae of hand-formatting. In return, Black gives you speed, determinism, and freedom from pycodestyle nagging about formatting. You will save time and mental energy for more important matters.  Blackened code looks the same regardless of the project you're reading. Formatting becomes transparent after a while and you can focus on the content instead. Black makes code review faster by producing the smallest diffs possible.

Also relevant, Ian Lee is an LC employee that used to be a major contributor to pycodestyle (# 2 in LOC), but AFAIK he now recommends people use black.

All that said, I'd be happy to use the other formatting tools, but I have no interest in configuring them or iterating on the format style.  So if we go with pycodestyle or yapf, I will just use whatever config is decided upon by others.

### LINKED ISSUE #21: Docstring Format
As discussed on the call today, we agreed on using Sphinx to generate our documentation from docstrings, but we need to decide/agree on a docstring format.  The three most popular formats are [RestructuredText](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html), [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html), and [Numpy-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy).

The latter two are supported in Sphinx via the [Napoleon extension](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/index.html) (which actually now ships as a part of Sphinx proper).  The Napoleon documentation has a great overview of the situation: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/index.html.  The tl;dr (too long didn't read) is: Napoleon pre-processes the google/numpy docstrings into the RestructuredText format before handing things over to sphinx for parsing and doc generation.  This means that all three formats can live together in the same codebase and even within the same docstring.  I'm not sure if google/numpy covers literally every feature of sphinx's docstrings, but if not, you can always fallback to ReST syntax for that particular feature/line.

-----

IMO, google and numpy style are way more legible (to me, the RestructuredText docstrings border on line noise sometimes with all the colons, poor grouping of type with description and poor separation of args, returns, exceptions, descriptions).  They are also much easier to remember and write than ReST.  IMO, the google format is easier to write than the numpy format since it uses tabs rather than `------`; so my vote is for the google style.  Ultimately, I would also be totally fine with the numpy format too.  Out of ease-of-use and legibility concerns, I vote against the ReST format/syntax.

### LINKED ISSUE #22: Type Annotations
Since we're talking about [code style](https://github.com/ExaWorks/psi-j-python/issues/20) and [docstring format](https://github.com/ExaWorks/psi-j-python/issues/21), I was wondering if it would be worth discussing type annotations. It sounds like everyone else is intending to annotate everything, which is fine by me, although I wonder if they might be optional until a given project (assuming that the policies put in place in this repo would carry over to Python code in other ExaWorks repos) finds that they would really be beneficial.

### PR DESCRIPTION
**Pull request by SteVwonder**
- Copy the CONTRIBUTING guide from the SDK repo
- add notes to CONTRIBUTING about the docstring format, type annotations, and code style
- add a make target for `autopep8`, add it to the dev-requirements.txt, and run it on the codebase
- fix `make docs` build due to docstring syntax errors

Closes #20
Closes #21
Closes #22

### CODE REVIEW

**Comment by hategan** [comment]
Looks great. Seems like there is not a lot of reformatting happening, so this is a pretty clean PR.

**Review by hategan — APPROVED** [review]

**Comment by SteVwonder** [comment]
Sorry. Just realized that my changes to the docstrings introducted a whitespace issue that `autopep8` didn't like.  Just force-pushed with the appropriate newline added.

**Review by jameshcorbett — APPROVED** [review]
Looks good to me!

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+- [Create your own
+  fork](https://docs.github.com/en/github/getting-started-with-github/fork-a-repo).
+- When you are starting a new set of changes, [fetch any changes from the
+  upstream
+  repo](https://matplotlib.org/stable/devel/gitwash/development_workflow.html#update-the-mirror-of-trunk),
+  and [start a new feature branch on your fork from
+  that](https://matplotlib.org/stable/devel/gitwash/development_workflow.html#make-a-new-feature-branch).
+- Make a new branch for each separable set of changes — ["one task, one
+  branch"](https://mail.python.org/pipermail/ipython-dev/2010-October/005632.html)
+- [Each commit should make one change](https://dev.to/ruanbrandao/how-to-make-good-git-commits-256k).
```
Comment (mturilli) [inline CONTRIBUTING.md:23]: Eliminate the period after `change`?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  to aide reviewing and (in the worst case) simplify reverting it in the future.
+  - A patch commit message should consist of a single short (less than 50
+    character) line summarizing the change, optionally followed by a blank line
```
Comment (mturilli) [inline CONTRIBUTING.md:26]: [...] single short [...] `sentence`?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    and then a more thorough description.
+  - Where applicable, a PR or commit message body should reference an Issue by
+    number (e.g. `Fixes #33`).
+- If you can possibly avoid it, avoid merging upstream branches or any other
```
Comment (mturilli) [inline CONTRIBUTING.md:30]: `If you can possibly avoid it,` -> If possible?

**Review by mturilli — APPROVED** [review]
Left few minor comments, please feel free to ignore them if not useful. Otherwise, LGTM.

**Review by andre-merzky — APPROVED** [review]
LGTM.  I'll not push the button just yet as there are comments from @mturilli.

File: CONTRIBUTING.md (hunk 1 above)
Comment (SteVwonder) [reply CONTRIBUTING.md:23]: Good catch.  Looks like all but the line above use periods at the end.  Are you OK if I add a period to the above line for consistency and minimal changes?  Or should I remove periods at the end of every line?

File: CONTRIBUTING.md (hunk 3 above)
Comment (SteVwonder) [reply CONTRIBUTING.md:30]: :+1:

File: CONTRIBUTING.md (hunk 2 above)
Comment (SteVwonder) [reply CONTRIBUTING.md:26]: 👍

**Comment by SteVwonder** [comment]
Address @mturilli's comments, rebased, added a CI fix, and opened an issue for the remaining CI failures (#29 ).  I think this PR is good to go in :)

**Pull request merged**